File: 0templates/py/unionfind-r.py
class UF:
    def __init__(self, n):
        self.par = list(range(n))
        self.rank = [0] * n
        
    def find(self, query):
        if query == self.par[query]:
            return query
        u_par = self.find(self.par[query])
        self.par[query] = u_par
        return u_par

    def union(self, node1, node2):
        par1, par2 = self.find(node1), self.find(node2)
        if par1 == par2:
            return False

        if self.rank[par1] > self.rank[par2]:
            self.par[par2] = self.par[par1]
        elif self.rank[par2] > self.rank[par1]:
            self.par[par1] = self.par[par2]
        else:
            self.par[par1] = self.par[par2]
            self.rank[par2] += 1

        return True

File: !templates/cpp/unionfind-r.cpp
class UF
{
private:
    vector<int> par;
    vector<int> rank;
public:
    UF(int n)
    {
        par.resize(n);
        for (int i = 0; i < n; ++i)
        {
            par[i] = i;
        }
        rank.resize(n, 1);
    }
    int find(int x)
    {
        int p = par[x];
        while (p != par[p])
        {
            par[p] = par[par[p]];
            p = par[p];
        }
        return p;
    }
    void unionn(int x1, int x2)
    {
        int p1 = find(x1), p2 = find(x2);
        if (p1 == p2) return;
        if (rank[p1] > rank[p2])
        {
            par[p2] = par[p1];
        }
        else if (rank[p1] < rank[p2])
        {
            par[p1] = par[p2];
        }
        else
        {
            par[p1] = par[p2];
            ++rank[p2];
        }
    }
};
